### ISSUE TITLE
UnboundLocalError: local variable 'ipaddr' referenced before assignment

### ISSUE THREAD

**Issue by brokoler** [derek-shnosh/ztp-watcher#4]
Hello,

I followed your tutorial and installed ZTP-Watcher including all dependencies (some not listed in your guide) successfully.
In FreeZTP, I created custom keystore values and installed your script including all needed dependencies successfully.

The ZTP-Watcher service is running successfully, but when a merged configuration is discovered the script stops with an error message:

![image](https://user-images.githubusercontent.com/57031979/137125198-4e692405-6d30-4599-8a08-85e1c656b77f.png)

> Oct 13 13:28:07 ztp bash[799]:     self.run()
> Oct 13 13:28:07 ztp bash[799]:   File "/usr/local/lib/python3.8/dist-packages/watchdog/observers/api.py", line 199, in run
> Oct 13 13:28:07 ztp bash[799]:     self.dispatch_events(self.event_queue, self.timeout)
> Oct 13 13:28:07 ztp bash[799]:   File "/usr/local/lib/python3.8/dist-packages/watchdog/observers/api.py", line 372, in dispatch_events
> Oct 13 13:28:07 ztp bash[799]:     handler.dispatch(event)
> Oct 13 13:28:07 ztp bash[799]:   File "/usr/local/lib/python3.8/dist-packages/watchdog/events.py", line 272, in dispatch
> Oct 13 13:28:07 ztp bash[799]:     {
> Oct 13 13:28:07 ztp bash[799]:   File "ztp-watcher.py", line 76, in on_created
> Oct 13 13:28:07 ztp bash[799]:     x = threading.Thread(target=self.test_ssh, args=(hostname, hostaddr, ipaddr))
> Oct 13 13:28:07 ztp bash[799]: **UnboundLocalError: local variable 'ipaddr' referenced before assignment**

Assigning a static IP in VLAN 1 via my merged template didn't change the behavior.
I don't know where the error message is coming from, could this be a reference issue in the script?

The merged configuration is saved at /etc/ztp/logs/merged with following name:

> C3650_10.6.13.240_SERIAL_merged.cfg

It would be nice to get some help.

Best regards

**Comment by derek-shnosh** [comment]
What is your `ssh_method` as configured in the **ztpconfig.yaml** file?

**Comment by derek-shnosh** [comment]
@eggplanting - I think I see the issue here. When I was working on this I was primarily focused on the **parse** `ssh_method` and likely overlooked the `ipaddr` variable for other `ssh_method`s.

I just pushed a quick fix (0ab73b6d94dd9324491d2be15e300a2fe6d37c13), please pull it down and provide an update.

**Comment by eggplanting** [comment]
Hello @derek-shnosh 

I tested your latest code which results in following error:
![image](https://user-images.githubusercontent.com/57031979/137484704-422e3d57-c0fe-47c6-8d14-a49f87ae49aa.png)

Everything was installed from scratch and I followed your tutorial exactly.
Sadly I don't know how to fix this by myself, would be nice to see an update.

Thanks a lot and best regards

**Comment by eggplanting** [comment]
Hi @derek-shnosh 

thanks a lot for updating, it seems to work now.

When doing a fresh install I also had to create the folder for the log files first with following command:
`mkdir /etc/ztp/logs/merged`

Maybe you could add this command to your setup guide.

**Comment by brokoler** [comment]
Hi @derek-shnosh 

my switch was just configured via FreeZTP and your script discovered my merged configuration file.
Following error is occuring:

![image](https://user-images.githubusercontent.com/57031979/137498978-cb46ac97-ceb2-41c2-8f9d-dfa98f8b3b65.png)

ztpwatcher.log shows no error, but the firmware upgrade isn't initiated:

> 2021/10/15 03:38:14 PM -- OK - Configuration file found (ztpconfig.yaml).
> 2021/10/15 03:38:14 PM -- WARNING: Post provisioning configuration file not found (/var/git/ztp-watcher/postcfg.txt).
> 2021/10/15 03:38:14 PM -- ZTP Watcher fully loaded and running.
> 2021/10/15 05:48:38 PM -- OK - Configuration file found (ztpconfig.yaml).
> 2021/10/15 05:48:38 PM -- WARNING: Post provisioning configuration file not found (/var/git/ztp-watcher/postcfg.txt).
> 2021/10/15 05:48:38 PM -- ZTP Watcher fully loaded and running.
> 2021/10/15 05:49:55 PM -- OK - Configuration file found (ztpconfig.yaml).
> 2021/10/15 05:49:55 PM -- WARNING: Post provisioning configuration file not found (/var/git/ztp-watcher/postcfg.txt).
> 2021/10/15 05:49:55 PM -- ZTP Watcher fully loaded and running.
> 2021/10/15 03:52:26 PM -- New file detected: C3650_10.6.13.240_SERIAL_merged.cfg
> 2021/10/15 03:52:26 PM -- C3650: Verifying SSH reachability to 10.6.13.240 in 10s.
> 2021/10/15 03:52:43 PM -- C3650: SSH reachability verified after 2 attempt(s).

Still not working right now, sadly.

**Comment by derek-shnosh** [comment]
@eggplanting - based on the error, it seems possible that Nornir 3.x made some changes regarding how hosts can be passed (I wrote this using Nornir 2.x). To be honest, not sure Nornir is necessary here, I could probably get along just fine with only Netmiko.

I'll look into a quick fix for passing the host to Nornir for now, and possibly look at restructuring to use Netmiko moving forward.

This is a fairly broken module at this point... but in the meantime, if you're willing, would you mind uninstalling Nornir 3.x and installing Nornir 2.5.0 to test?

```bash
pip uninstall nornir
pip uninstall nornir-netmiko
pip install nornir==2.5.0
```

You may need to muck with the `import`s, but it seems you figured that out [here](https://github.com/helpdeskdan/ztp-watcher/issues/1).

**Comment by brokoler** [comment]
@derek-shnosh - Awesome, the service seems to work now. There was a second dependency when installing nornir==2.5.0, I had to downgrade ncclient too in following order:

`pip install ncclient==0.6.9`
`pip install nornir==2.5.0`

ztpwatcher.log shows following:
![image](https://user-images.githubusercontent.com/57031979/137524111-2a8ba4ea-cbb3-4af5-a8c6-d73b4fa96fd7.png)

But still transfering the image via TFTP didn't succeed, the image was not added to my flash: storage and it reloaded with it's old version.

Any tips how to troubleshoot this? My settings are looking good:
![image](https://user-images.githubusercontent.com/57031979/137523837-079e7aae-d9cc-4723-b972-6c7b18454343.png)

**Comment by derek-shnosh** [comment]
@eggplanting - let'd do some _print-debugging_. Please put the following on line # 182 of **ztp-watcher.py**, make sure to restart your service with `systemctl restart ztp-watcher` afterwards.

```python
Logger(copystatus)
```

Should look like this...

```python
            if '%Error' not in copystatus:
                Logger(copystatus)
                Logger(f'{hostname}: Image transfer completed after {copyduration}s -> set boot variable.')
                sw_log('Image transfer complete -> set boot variable.')
                # Set boot variable on switch.
                ...
```

Run through the process again and paste the log here.

**Comment by brokoler** [comment]
Transfering the image manually via following switch command is working fine, so it seems to be an issue of the script:
`copy tftp:cat3k_caa-universalk9.16.12.06.SPA.bin flash:`

Gonna test your debugging commands on weekend, I'm having trouble with the TFTP transfer speed even manually.
With blocksize 8192 and FreeZTP tftp server it takes more than 40 minutes to transfer my image. I don't know why.

Edit: Decreasing the blocksize to 512 actually increases transfer speeds. Not an issue related to this one though. Sorry.

Edit 2: issue is the TFTP server included with FreeZTP, I'm getting timeouts during transfer with a blocksize absove 512. Using a TFTP server on a different machine fixes the issue.

Edit 3: Using an external TFTP now, this one works great with a blocksize of 8192 bytes. Took around 8 minutes for a 450 MB file.

I'm going to proceed with the initial issue.

**Comment by brokoler** [comment]
@derek-shnosh thanks for your patience, please take a look at the new log output:

![image](https://user-images.githubusercontent.com/57031979/137553854-196f1a9e-c7bb-4fa9-a895-c6e6c7ce277a.png)

Serial and hostname is equal in my case, it's censored in the screenshot.

**Comment by derek-shnosh** [comment]
@eggplanting - ahh, try putting `file prompt quiet` in your switch config/template. Nornir/Netmiko is expecting the transfer to start or a `hostname#` prompt, but is instead getting confirmation prompts you usually get when doing copies (check [this page](https://blog.ipspace.net/2006/11/reduce-noise-generated-by-cisco-ios.html) for details).

**Comment by derek-shnosh** [comment]
> Edit: Decreasing the blocksize to 512 actually increases transfer speeds. Not an issue related to this one though. Sorry.

@eggplanting - I think you might be misinterpreting the `!` markers appearing _faster_ with the blocksize set to 512 as the transfer actually being faster. The `!` markers appear per block copied, so when copying at 512 bytes, they'll appear faster than with the blocksize set to 8192. However having the block size set to 8192 will ultimately accomplish the transfer faster, at least in my experimentation.

Check out this [reference article](https://www.cisco.com/c/en/us/support/docs/wireless/5760-wireless-lan-controller/117636-technote-tftpfile-00.html) from Cisco.

**Comment by brokoler** [comment]
@derek-shnosh It's finally working, thank you so much! It was a pleasure debugging the issue with you.

Regarding TFTP transfer speed: It's only working fast with my external TFTP server (Windows with TFTP64)
Via FreeZTP the tftp transfer was looking like this: `!.....!.....!...!`
It seemed really unstable and was only working when blocksize was decreased to 512, but really slow then.

I'm fine with the external TFTP server, no problem here. Updating takes stable 8 minutes now. :)

Would it be possible to add an "auto reboot after firmware" function in the settings of ztp-watcher?
It would allow to decide between rebooting manually or automatically when the update was completely transfered.

Such a great project, it really enhances FreeZTP a lot.

**Comment by derek-shnosh** [comment]
Glad to hear things are working now and thanks for reporting the issue.

I will add the auto-reboot feature request to my notes for whenever I get around to re-addressing this code. Full disclosure, not exactly sure when that'll be. 😉

Being that you're _off to the races_, so to speak, I'm going to close this issue out. Feel free to reopen it if you need to.

**Comment by brokoler** [comment]
@derek-shnosh 

Aww, sadly it didn't set the boot variable for the switch.
TFTP transfer copied the file to the switch, but the log still tells me it only took 11 seconds (instead 8 minutes) and the boot variable was set.

Checking on the switch, the file was transfered to flash but no change to the boot variable. Nothing special in the logs.
Could you tell me which line to edit in the .py file to further debug?

`show boot` shows me `flash:packages.conf`, my IOS is running in INSTALL mode. Nothing was changed by ztpwatcher.
Is INSTALL mode even supported or is it BUNDLE mode only?

**Comment by derek-shnosh** [comment]
It's interesting to me that it thinks the copy completes after 11s... the script was written to be used with BUNDLE mode installations; it's admittedly a bit dated right now and likely needs a substantial rewrite, which unfortunately won't be anything that I can address for the foreseeable future.

TLDR; the script, in its current form, will not perform a proper upgrade in INSTALL mode.

**Comment by eggplanting** [comment]
Thank you, understood.

Could you maybe give me a short instruction how to use EEM scripts with FreeZTP? I don't understand where to put them.
https://github.com/PackeTsar/freeztp/blob/latest/TIPS.md#use-case-multi-platform-ios--ios-xe-upgrade

I'm trying to use following EEM scripts, which should allow install mode.
